packages/react-components/src/components/xy-output-component.tsx

```diff
@@ -288,6 +291,191 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
         document.addEventListener('mouseup', this.endSelection);
     }
 
+    private updateRange(rangeStart: number, rangeEnd: number,percentMousePosition: number) {
+        if (rangeEnd < rangeStart) {
+            const temp = rangeStart;
+            rangeStart = rangeEnd;
+            rangeEnd = temp;
+        }
+        this.props.unitController.viewRange = {
+            start: rangeStart,
+            end: rangeEnd
+        };
+        this.updatePosition(percentMousePosition);
+    }
+
+    private updatePosition(percentMousePosition: number) {
+        const newPosition = percentMousePosition * this.props.unitController.viewRangeLength;
+        this.positionZoomMouse = newPosition + this.props.unitController.viewRange.start;
+    }
+
+    private newRangeZoom(zoomIn: number, startPosition: number, endPosition: number): number[] {
+        let positionZoom = 0;
+        if (this.isZoomCenter) {
+            positionZoom = this.positionZoomCenter;
+        } else {
+            positionZoom = this.positionZoomMouse;
+        }
+        const newStartRange = positionZoom - (startPosition - zoomIn * (startPosition * TEN_PERCENT));
+        const newEndRange = positionZoom + (endPosition - zoomIn * (endPosition * TEN_PERCENT));
+        return [newStartRange, newEndRange];
+    }
+
+    private zoomIn(startPos: number, endPos: number, percentMousePosition: number) {
+        const zoom = 1;
+        let startPosition = 0;
+        let endPosition = 0;
+        if (this.mouseIsDown) {
+            startPosition = startPos;
+            endPosition = endPos;
+        } else {
+            startPosition =   this.newRangeZoom(zoom, startPos, endPos)[0];
+            endPosition = this.newRangeZoom(zoom, startPos, endPos)[1];
+        }
+        this.updateRange(startPosition, endPosition, percentMousePosition);
+    }
+
+    private zoomOut(startPosition: number, endPosition: number, percentMousePosition: number) {
+        const zoom = -1;
+        this.updateRange( this.newRangeZoom(zoom, startPosition, endPosition)[0], this.newRangeZoom(zoom, startPosition, endPosition)[1],percentMousePosition);
+    }
+
+    private NewstartEndRangePanning(shift: number, percentRange: number): number[] {
+        const startRange = this.props.unitController.viewRange.start + (shift * percentRange);
+        const endRange = this.props.unitController.viewRange.end + (shift * percentRange);
+        const arr = [startRange, endRange];
+        return arr;
+    }
+
+    private updatePositionAfterPann(right: number, distance: number){
+        this.positionZoomCenter = this.positionZoomCenter + right * distance;
+        this.positionZoomMouse = this.positionZoomMouse + right * distance;
+    }
+
+    private shift(right: number, distance: number) {
+        const percentRange = this.props.unitController.viewRangeLength * TEN_PERCENT;
+        if (this.NewstartEndRangePanning(right, percentRange)[0] < 0) {
```

Comment:
I would just calculate the expected new range once and handle the limits, not call the same method repeatedly.